### DISCUSSION TITLE
Does the MeshCentral API not report powered on or connected status for Windows agents?

### DISCUSSION THREAD

**Discussion by DaanSelen** [Ylianst/MeshCentral#6643] [General]
As the title asks, does it not?

```
"mesh//lgNwoUxCblablabla57ghcd9": [
            {
                "mtype": 2,
                "_id": "nblablablaCft1AjX@",
                "icon": 1,
                "name": "N-DESKTOP1",
                "rname": "N-DESKTOP1",
                "domain": "",
                "host": null,
                "ip": "172.16.25.234",
                "osdesc": "Microsoft Windows 11 Home - 24H2/26100",
                "av": [
                    {
                        "product": "Windows Defender",
                        "updated": true,
                        "enabled": true
                    }
                ],
                "wsc": {
                    "antiVirus": "OK",
                    "autoUpdate": "OK",
                    "firewall": "OK"
                },
                "users": [],
                "lastbootuptime": 1734779438000,
                "defender": {
                    "RealTimeProtection": true,
                    "TamperProtected": true
                }
            }
        ]
```

Compared to a Linux machine it does not report the powered on or connected status?

**Comment by si458** [comment]
Well what api command did u call to get the current information as it might be a different command

**Reply by DaanSelen** [reply]
Both getDeviceDetails and Nodes.
In both I do not see a direct pwr or similar for Windows agents.

**Reply by si458** [reply]
so its a mix of different things, we dont have a value like 'online' or 'offline' as such
you should have a value called `conn` and if its there and set to 1 or above, then the device is online
if its missing or set as 0 then its offline

**Reply by DaanSelen** [reply]
That's weird...

```shell
ethernet', 'status': 'up', 'address': 'fe80::640c:51ff:feb7:da00%6', 'family': 'IPv6', 'netmask': 'ffff:ffff:ffff:ffff::'}]}, 'updateTime': 1735337322937, '_id': 'ifnode//1X', 'domain': '', 'type': 'ifinfo'}}, created_at=None, lastaddr=None, lastconnect=None connected=False, powered_on=False, os_description='Debian GNU/Linux 12 (bookworm)', links={}, **{'type': 'node', '_id': 'node//6LwE5CBd2VGOngoOx$Ox9kC5ZwqBRn6Kbf3E630myp4fJ1V0hRPjW@wwLqS41b1X', 'meshid': 'mesh//QRDp1u', 'lastbootuptime': 1734043786000
```

This is a snippit of a machine that for sure is online... 100% and it reports as false...

It's one of these:

![image](https://github.com/user-attachments/assets/f494f9b2-85cb-4af1-bdcf-885fa5066fe6)

**Reply by si458** [reply]
I will double check later as I'm not 100% sure

**Reply by si458** [reply]
so ive just tested it here `{ action: 'nodes', responseid: 'meshctrl' }` and its returning all the devices,
and each device that is online shows `pwr` as 1 IF its online
and `conn` is either 1 for agent or 2 for amt, again IF the device is online

are you talking directly with websocket or a library of come kind?

**Reply by DaanSelen** [reply]
Hmmm I was using the Python implementation, but it just sends: `{ action: 'getDeviceDetails', type: 'json'  }`.

I will also double check.

**Reply by si458** [reply]
`getDeviceDetails` only gets you details about the device, it wont show you the status of the devce

**Reply by DaanSelen** [reply]
Okay... the `getDeviceDetails` does NOT report it, but the normal call `nodes` does. Is it not counterintuitive?
Should it not tell the detail if the device is connected?

**Reply by si458** [reply]
in my eyes no,
because `getDeviceDetails` would return me details about the device (think the details tab)
where as the `nodes` would get me a list of devices and their statuses (think the listing page)
i can ask @Ylianst for his opinion on the matter tho,
then maybe we could add in the `pwr` and `conn` into `getDeviceDetails`

**Reply by si458** [reply]
~~P.S you can use the `--filterid [id]` with `nodes` to only return the single device you are asking for~~
scratch that, this is a function in meshctrl.js and not the websocket itself

**Reply by DaanSelen** [reply]
Yes lets leave it at that. For me personally it would make it easier, but let me know what he thinks. If he does agree, then good. If not I'll make another call for the nodes.

**Reply by DaanSelen** [reply]
> P.S you can use the `--filterid [id]` with `nodes` to only return the single device you are asking for

This would make another call for every device reported. I think its more efficient to make 2 general calls and just reference the data.

**Reply by si458** [reply]
@DaanSelen the way the web ui does it
1. it calls `nodes` to list all the devices, 
2. then when u pick a device, it calls other functions to get details about the device (`deviceinfo` for meshctrl.js example)
```
ws.send(JSON.stringify({ action: 'getnetworkinfo', nodeid: args.id, responseid: 'meshctrl' }));
ws.send(JSON.stringify({ action: 'lastconnect', nodeid: args.id, responseid: 'meshctrl' }));
ws.send(JSON.stringify({ action: 'getsysinfo', nodeid: args.id, nodeinfo: true, responseid: 'meshctrl' }));
```
the idea of `getDeviceDetails` from the looks of it is to merge those 3 commands together so its a single call

you can also specify `nodeids` with a list of devices you want it to return data for

example: `{ action: 'getDeviceDetails', nodeids: ['id1','id2'], responseid: 'meshctrl' }`

**Reply by si458** [reply]
OMG your issue about merging the `pwr,conn` was already asked/answered 3 years ago 😆 
https://github.com/Ylianst/MeshCentral/issues/2933

edit: it appears to have maybe been missed! so i might add it in for you!
can you open an enhancement request plz asking for this feature 👍

**Reply by DaanSelen** [reply]
> OMG your issue about merging the `pwr,conn` was already asked/answered 3 years ago 😆 #2933
> 
> edit: it appears to have maybe been missed! so i might add it in for you! can you open an enhancement request plz asking for this feature 👍

I've read the Issue, and looks like Ylianst forgot about it. I don't blame him! Lets pick it up if he agrees.
